torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1544,10 +1537,20 @@ def _full_post_state_dict_hook(
         if not state_dict:
             return state_dict
 
+        offload_to_cpu = self._state_dict_config.offload_to_cpu
+        cpu_device = torch.device("cpu")
         for key in state_dict:
+            clean_key = clean_tensor_name(key)
+            # Do not need to clone buffers
+            if clean_key in self._buffer_names:
+                # Offload the buffer to CPU if needed since it is not done in
+                # `_summon_full_params()`
```

Comment:
should we eventually push this down to `summon_full_params` so that we get it for free? Feel free to file an issue if you feel that would be better than doing it manually here. It may also make things more consistent with `summon_full_params`.

Reply:
It would be clean if we could do this in `summon_full_params()`, but in my understanding, there is one main barring factor:
If we do it in `summon_full_params()`, then we need to destructively modify the actual buffer variable's data to point to CPU storage before the `state_dict()` logic triggers and restore the data to point back to GPU storage after the `state_dict` has been computed. Without some care, the original GPU storage will get freed and later re-allocated, which is wasteful.

I chose to do it in the full post state dict hook since then we only incur the necessary cost of allocating CPU storage for the buffer's data. The original buffer's data pointing to GPU storage is not freed and hence does not need to be reallocated.